Zend/zend_dimension_handlers.h

```diff
@@ -0,0 +1,46 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Author: Gina Peter Banyard <girgias@php.net>                         |
+   +----------------------------------------------------------------------+
+*/
+
+/* This really should not be needed... */
+#ifndef ZEND_DIMENSION_HANDLERS_H
+#define ZEND_DIMENSION_HANDLERS_H
+
+//BEGIN_EXTERN_C()
+
+#include "zend_types.h"
+
+typedef struct _zend_class_dimensions_functions {
+	/* rv is a slot provided by the callee that is returned */
+	zval *(*/* const */ read_dimension)(zend_object *object, zval *offset, zval *rv);
+	bool  (*/* const */ has_dimension)(zend_object *object, zval *offset);
+	zval *(*/* const */ fetch_dimension)(zend_object *object, zval *offset, zval *rv);
```

Comment:
I'm not understanding something. Why do we pass in a `zval *rv`, which I assume is the "return value", but then also return it? Wouldn't it be better to pass in the return value, and return `void` or `zend_result` instead? Are they separate zvals in some cases other than an error?

Reply:
Yes, there are seperate zvals in some cases, in which case we need to make the returned zval IS_INDIRECT. The lact of *rv as an arg and return value has been an issue with properly supporting fetches for the property hook RFC as the get_ptr_ptr handler doesn't have this mechanism.